flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/reuse/MultipleInputNodeCreationProcessor.java

```diff
@@ -0,0 +1,481 @@
+				Preconditions.checkState(
+					wrapper == group.root,
+					"The only member of a multiple input group is not its root. This is a bug.");
+				// optimization 1. we clean up multiple input groups with only 1 member,
+				// unless one of its input is a FLIP-27 source (for maximizing source chaining),
+				// however unions do not apply to this optimization because they're not real operators
+				if (isUnion || wrapper.inputs.stream().noneMatch(inputWrapper -> isNewSource(inputWrapper.execNode))) {
+					wrapper.removeFromGroup();
+				}
+				continue;
+			}
+
+			if (!isTailOfMultipleInputGroup(wrapper)) {
+				// we're not removing a node from the middle of a multiple input group
+				continue;
+			}
+
+			boolean shouldRemove = false;
+			if (isUnion) {
+				// optimization 2. we do not allow union to be the tail of a multiple input
+				// as we're paying extra function calls for this, unless one of the united
+				// input is a FLIP-27 source
+				shouldRemove = wrapper.inputs.stream().noneMatch(inputWrapper -> isNewSource(inputWrapper.execNode));
+			} else if (wrapper.inputs.size() == 1) {
+				// optimization 3. for one-input operators we'll remove it unless its input
+				// is an exchange or a FLIP-27 source, this is mainly to avoid the following
+				// pattern:
+				// non-chainable source -> calc --\
+				//                                 join ->
+				// non-chainable source -> calc --/
+				// if we move two calcs into the multiple input group rooted at the join, we're
+				// directly shuffling large amount of records from the source without filtering
+				// by the calc
+				ExecNode<?, ?> input = wrapper.inputs.get(0).execNode;
+				shouldRemove = !(input instanceof BatchExecExchange) &&
+					!(input instanceof StreamExecExchange) &&
+					!isNewSource(input);
+			}
+
+			// optimization 4. for singleton operations (for example singleton global agg)
+			// we're not including it into the multiple input node as we have to ensure that
+			// the whole multiple input can only have 1 parallelism.
+			// continuous singleton operations connected by forwarding shuffle will be dealt
+			// together with optimization 3
+			shouldRemove |= wrapper.inputs.stream().anyMatch(inputWrapper ->
+				inputWrapper.execNode instanceof BatchExecExchange &&
+					((BatchExecExchange) inputWrapper.execNode)
+						.distribution.getType() == RelDistribution.Type.SINGLETON);
+
+			if (shouldRemove) {
+				wrapper.removeFromGroup();
+			}
+		}
+	}
+
+	private boolean isTailOfMultipleInputGroup(ExecNodeWrapper wrapper) {
+		Preconditions.checkNotNull(
+			wrapper.group,
+			"Exec node wrapper does not have a multiple input group. This is a bug.");
+		for (ExecNodeWrapper inputWrapper : wrapper.inputs) {
+			if (inputWrapper.group == wrapper.group) {
+				// one of the input is in the same group, so this node is not the tail of the group
+				return false;
+			}
+		}
+		return true;
+	}
+
+	private boolean isNewSource(ExecNode<?, ?> node) {
```

Comment:
`SourceProvider` could also provide new `Source`

Reply:
FLIP-146 is not ready when this PR is submitted. I'll rebase the master branch.